book/html.md

```diff
@@ -5,408 +5,579 @@ prev: text
+list of tokens. We added:
+
+- A parser to transform HTML tokens to a tree
+- Layout operating recursively on the tree
+- Code to recognize and handle attributes on elements
+- Automatic fixes for some malformed HTML documents
+
+The tree structure of HTML is essential to display visually complex
+web pages, as we will see in the [next chapter](layout.md).
+
+::: {.signup}
+:::
 
 
 Exercises
 =========
 
--   HTML documents almost always start with \"doctype declarations\". These
-    look like tags that begin with `!doctype` and don\'t have close
-    tags. Doctype declarations, when they are present, are always the
-    first token in a document. Skip doctype declarations in the parser.
--   Update the HTML lexer and parser to support comments. Comments in
-    HTML begin with `<!--` and end with `-->`. However, comments aren\'t
-    the same as tags: they can contain any text, including open or close
-    tags. Comments should be a new token type, which `parse` should
-    ignore.
--   Update the HTML lexer or parser to support *entities*. Entities in
-    HTML begin an ampersand `&`, then contain letters and numbers, and
-    end with a semicolon \"`;`\"; they resolve to particular characters.
-    Implement support for `&amp;`, `&lt;`, `&gt;`, and `&quot;`, which
-    expand to ampersand, less than, greater than, and the quotation
-    mark. Should you handle entities in the lexer, the parser, or in an
-    earlier pass? Consider that attribute values can contain entities.
--   For some tags, it doesn\'t make sense to have one inside the other.
-    For example, it\'s not clear what it would mean for one paragraph to
-    contain another, so the most common reason for this to happen in a
-    web page is that someone forgot a close tag. Change the parser so
-    that a document like `<p>hello<p>world</p>` results in two sibling
-    nodes instead of one paragraph inside another.
--   The attribute parser doesn\'t correctly handle attribute values
-    that contain spaces, which is valid when the attribute is quoted.
-    Fix this case in the attribute parser. You will likely need to
-    loop over the attribute character-by-character.
-
-[^1]: To be clear: there is additional structure, we\'re just ignoring
-    it for now.
-
-[^2]: Tracking the parent like I\'m doing here is going to be useful for
-    parsing.
-
-[^3]: Yes, it\'s a crazy system, and for a few years in the early '00s
-    the W3C tried to [do away with it](https://www.w3.org/TR/xhtml1/).
-    They failed.
-
-[^4]: Some people put a slash at the end of a self-closing tag (like
-    `<br/>`) but they don\'t have to: `<br>` is self-closing with and
-    without that slash.
-
-[^5]: Where `href` stands for \"hypertext reference\".
-
-[^6]: If you implemented earlier exercises like support for `<a>`,
-    `<small>`, and `<big>`, you will likely have more parts to the
-    state.
+*Entities:* Implement support for `&amp;`, `&lt;`, `&gt;`, and
```

Comment:
Rework this paragraph to refer back to the exercise I added in the other chapter, and
make this exercise about adding the other entities not in the other exercise (if the
student had already done the other exercise).

Reply:
Great point. I ended up eliminating this exercise and replacing it with something I think is more interesting: turning your lexer into a syntax highlighter which you'd run for `view-source:` URLs.